File: pkg/daemon/spawn.go
package daemon

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"syscall"
	"time"
)

// SpawnOptions control how EnsureDaemon starts a daemon that isn't running.
type SpawnOptions struct {
	// NoSpawn fails with DAEMON_UNAVAILABLE instead of starting one.
	NoSpawn bool
	// IdleTimeout in seconds; 0 keeps the daemon's default, -1 disables.
	IdleTimeout int
	// HTTPAddr / Token are passed through to `serve`.
	HTTPAddr string
	Token    string
	// StartTimeout bounds the wait for READY (default 20s).
	StartTimeout time.Duration
	// ExtraArgs are appended to the serve command line.
	ExtraArgs []string
	// Log receives progress lines ("Starting daemon …"); nil = silent.
	Log func(format string, args ...any)
}

// Binary returns the executable used to spawn daemons: MAESTRO_DAEMON_BIN
// or the current binary.
func Binary() (string, error) {
	if v := os.Getenv(EnvBin); v != "" {
		return v, nil
	}
	return os.Executable()
}

// EnsureDaemon returns a client for the named daemon, spawning it in the
// background when it isn't running. started reports whether this call
// spawned it.
func EnsureDaemon(ctx context.Context, name string, opts SpawnOptions) (c *Client, started bool, err error) {
	if err := ValidateName(name); err != nil {
		return nil, false, err
	}
	if c, err := Dial(name); err != nil {
		return nil, false, err
	} else if c != nil {
		if perr := c.Ping(3 * time.Second); perr == nil {
			return c, false, nil
		} else if opts.Log != nil {
			opts.Log("daemon %q (pid %d) is not answering: %v", name, c.Info.PID, perr)
		}
		// Recorded pid alive but socket dead: treat as stale.
		if !PIDAlive(c.Info.PID) {
			RemoveRunFiles(name)
		} else {
			return nil, false, Errorf(CodeDaemonUnavailable,
				"daemon %q (pid %d) exists but does not answer on %s; stop it with `maestro-daemon stop --name %s`",
				name, c.Info.PID, c.Info.SocketPath, name)
		}
	}
	if opts.NoSpawn {
		return nil, false, Errorf(CodeDaemonUnavailable, "daemon %q is not running", name)
	}
	c, err = spawn(ctx, name, opts)
	if err != nil {
		return nil, false, err
	}
	return c, true, nil
}

func spawn(ctx context.Context, name string, opts SpawnOptions) (*Client, error) {
	bin, err := Binary()
	if err != nil {
		return nil, WrapErr(CodeDaemonUnavailable, err)
	}
	if err := os.MkdirAll(RunDir(name), 0o700); err != nil {
		return nil, WrapErr(CodeDaemonUnavailable, err)
	}
	logPath := StartupLogPath(name)
	logFile, err := os.OpenFile(logPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0o600)
	if err != nil {
		return nil, WrapErr(CodeDaemonUnavailable, err)
	}
	defer logFile.Close()

	args := []string{"serve", "--name", name}
	switch {
	case opts.IdleTimeout < 0:
		args = append(args, "--idle-timeout", "0")
	case opts.IdleTimeout > 0:
		args = append(args, "--idle-timeout", strconv.Itoa(opts.IdleTimeout))
	}
	if opts.HTTPAddr != "" {
		args = append(args, "--http", opts.HTTPAddr)
	}
	if opts.Token != "" {
		args = append(args, "--token", opts.Token)
	}
	args = append(args, opts.ExtraArgs...)

	cmd := exec.Command(bin, args...)
	cmd.Stdout = logFile
	cmd.Stderr = logFile
	cmd.Stdin = nil
	cmd.Env = os.Environ()
	// Own session so a Ctrl-C in the caller's terminal doesn't kill it.
	cmd.SysProcAttr = &syscall.SysProcAttr{Setsid: true}
	if opts.Log != nil {
		opts.Log("Starting daemon %q: %s %s", name, bin, strings.Join(args, " "))
	}
	if err := cmd.Start(); err != nil {
		return nil, Errorf(CodeDaemonUnavailable, "spawn %s: %v", bin, err)
	}
	pid := cmd.Process.Pid
	exited := make(chan error, 1)
	go func() { exited <- cmd.Wait() }()

	timeout := opts.StartTimeout
	if timeout <= 0 {
		timeout = 20 * time.Second
	}
	deadline := time.Now().Add(timeout)
	for {
		select {
		case werr := <-exited:
			return nil, Errorf(CodeDaemonUnavailable, "daemon %q (pid %d) exited during startup: %v\n%s",
				name, pid, werr, tailFile(logPath, 20))
		case <-ctx.Done():
			_ = cmd.Process.Kill()
			return nil, WrapErr(CodeInterrupted, ctx.Err())
		case <-time.After(100 * time.Millisecond):
		}
		if info, _ := ReadInfo(name); info != nil && info.PID == pid {
			c := NewUnixClient(info.SocketPath)
			c.Name = name
			c.Info = info
			if err := c.Ping(2 * time.Second); err == nil {
				if opts.Log != nil {
					opts.Log("Daemon %q ready (pid %d, socket %s)", name, pid, info.SocketPath)
				}
				return c, nil
			}
		}
		if time.Now().After(deadline) {
			_ = cmd.Process.Kill()
			return nil, Errorf(CodeDaemonUnavailable, "daemon %q (pid %d) did not become ready within %v\n%s",
				name, pid, timeout, tailFile(logPath, 20))
		}
	}
}

// WaitForExit blocks until the daemon's pid is gone or timeout passes.
func WaitForExit(info *Info, timeout time.Duration) bool {
	deadline := time.Now().Add(timeout)
	for PIDAlive(info.PID) {
		if time.Now().After(deadline) {
			return false
		}
		time.Sleep(100 * time.Millisecond)
	}
	return true
}

// tailFile returns the last n lines of a file for error messages.
func tailFile(path string, n int) string {
	data, err := os.ReadFile(path)
	if err != nil || len(data) == 0 {
		return ""
	}
	lines := strings.Split(strings.TrimRight(string(data), "\n"), "\n")
	if len(lines) > n {
		lines = lines[len(lines)-n:]
	}
	return fmt.Sprintf("--- %s ---\n%s", path, strings.Join(lines, "\n"))
}

File: npm/maestro-d/src/spawn.ts
// Starting a daemon that isn't running (mirrors pkg/daemon/spawn.go and the
// Edge CLI's ensureEngine): spawn `maestro-d serve --name <name>`
// detached with its output in startup.log, then poll daemon.json until the
// new process answers.

import { spawn } from 'node:child_process'
import * as fs from 'node:fs'

import { binaryPath } from './binary.js'
import { MaestroError } from './errors.js'
import { liveInfo, pidAlive, readInfo, removeRunFiles, runDir, startupLogPath, tailFile, validateName } from './runfiles.js'
import { Transport } from './transport.js'
import type { DaemonInfo } from './types.js'

export interface SpawnOptions {
  /** Fail with DAEMON_UNAVAILABLE instead of starting one. */
  spawn?: boolean
  /** Seconds of inactivity before the daemon exits; 0 = never. Default 1800. */
  idleTimeout?: number
  /** Also listen on TCP (`127.0.0.1:7788`). */
  http?: string
  /** Bearer token required on TCP. */
  token?: string
  /** Milliseconds to wait for the daemon to come up (default 20000). */
  startTimeout?: number
  /** Extra `serve` arguments. */
  extraArgs?: string[]
  /** Path of the maestro-d binary (else MAESTRO_D_BIN, bundled, PATH). */
  bin?: string
  /** Progress lines ("Starting daemon …"). */
  log?: (message: string) => void
  signal?: AbortSignal
}

export interface Ensured {
  info: DaemonInfo
  transport: Transport
  /** True when this call started the daemon. */
  started: boolean
}

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    const t = setTimeout(() => {
      signal?.removeEventListener('abort', onAbort)
      resolve()
    }, ms)
    function onAbort(): void {
      clearTimeout(t)
      reject(MaestroError.of('INTERRUPTED', 'aborted while waiting for the daemon'))
    }
    signal?.addEventListener('abort', onAbort, { once: true })
  })
}

async function ping(t: Transport, timeout: number): Promise<boolean> {
  try {
    await t.json('/v1/status', { timeout })
    return true
  } catch {
    return false
  }
}

/** Connect to the named daemon, spawning it when needed. */
export async function ensureDaemon(name: string, opts: SpawnOptions = {}): Promise<Ensured> {
  validateName(name)
  const existing = liveInfo(name)
  if (existing) {
    const t = new Transport({ socketPath: existing.socketPath })
    if (await ping(t, 3000)) {
      return { info: existing, transport: t, started: false }
    }
    t.close()
    opts.log?.(`daemon "${name}" (pid ${existing.pid}) is not answering`)
    if (!pidAlive(existing.pid)) {
      removeRunFiles(name)
    } else {
      throw MaestroError.of(
        'DAEMON_UNAVAILABLE',
        `daemon "${name}" (pid ${existing.pid}) exists but does not answer on ${existing.socketPath}; ` +
          `stop it with \`maestro-d stop --name ${name}\``,
      )
    }
  }
  if (opts.spawn === false) {
    throw MaestroError.of('DAEMON_UNAVAILABLE', `daemon "${name}" is not running`)
  }
  return spawnDaemon(name, opts)
}

async function spawnDaemon(name: string, opts: SpawnOptions): Promise<Ensured> {
  const bin = binaryPath(opts.bin)
  fs.mkdirSync(runDir(name), { recursive: true, mode: 0o700 })
  const logPath = startupLogPath(name)
  const logFd = fs.openSync(logPath, 'w', 0o600)

  const args = ['serve', '--name', name]
  if (opts.idleTimeout !== undefined) args.push('--idle-timeout', String(opts.idleTimeout))
  if (opts.http) args.push('--http', opts.http)
  if (opts.token) args.push('--token', opts.token)
  if (opts.extraArgs) args.push(...opts.extraArgs)

  opts.log?.(`Starting daemon "${name}": ${bin} ${args.join(' ')}`)
  let child
  try {
    // Own process group so a Ctrl-C in the caller's terminal does not kill it.
    child = spawn(bin, args, { detached: true, stdio: ['ignore', logFd, logFd], env: process.env })
  } catch (err) {
    fs.closeSync(logFd)
    throw MaestroError.of('DAEMON_UNAVAILABLE', `spawn ${bin}: ${(err as Error).message}`, { cause: err })
  }
  fs.closeSync(logFd)

  let exited: { code: number | null; signal: NodeJS.Signals | null } | undefined
  let spawnError: Error | undefined
  child.on('exit', (code, signal) => {
    exited = { code, signal }
  })
  child.on('error', (err) => {
    spawnError = err
  })
  child.unref()
  const pid = child.pid ?? 0

  const timeout = opts.startTimeout ?? 20_000
  const deadline = Date.now() + timeout
  for (;;) {
    await sleep(100, opts.signal)
    if (spawnError) {
      throw MaestroError.of('DAEMON_UNAVAILABLE', `spawn ${bin}: ${spawnError.message}`, { cause: spawnError })
    }
    if (exited) {
      const how = exited.signal ? `signal ${exited.signal}` : `exit ${exited.code}`
      throw MaestroError.of('DAEMON_UNAVAILABLE', `daemon "${name}" (pid ${pid}) exited during startup: ${how}\n${tailFile(logPath, 20)}`)
    }
    const info = readInfo(name)
    if (info && info.pid === pid) {
      const t = new Transport({ socketPath: info.socketPath })
      if (await ping(t, 2000)) {
        opts.log?.(`Daemon "${name}" ready (pid ${pid}, socket ${info.socketPath})`)
        return { info, transport: t, started: true }
      }
      t.close()
    }
    if (Date.now() > deadline) {
      try {
        child.kill()
      } catch {
        // already gone
      }
      throw MaestroError.of('DAEMON_UNAVAILABLE', `daemon "${name}" (pid ${pid}) did not become ready within ${timeout}ms\n${tailFile(logPath, 20)}`)
    }
  }
}

/** Wait until the daemon's pid is gone; false on timeout. */
export async function waitForExit(info: DaemonInfo, timeoutMs: number): Promise<boolean> {
  const deadline = Date.now() + timeoutMs
  while (pidAlive(info.pid)) {
    if (Date.now() > deadline) return false
    await sleep(100)
  }
  return true
}
